### ISSUE TITLE
utils/bind-data-test-attributes: Handle failing `set()` with assert instead of warning

### ISSUE THREAD

**Issue by mathieupoteriepeopledoc** [simplabs/ember-test-selectors#204]
[Reference Commit]( https://github.com/simplabs/ember-test-selectors/commit/91fdec7d0797bd32185d9a0c7a7da93884d3f5b0#diff-749f8a6a3957ed97380d08e488fd3cb)
After experiencing some issues after trying to add a data-test-selector to tagless components, I realized afterwards, by looking at the commit on master, that a warning was created to give some information. This is a great idea, but maybe not explicit enough (I didn't see this warning, in the middle of other warnings). 
If you agree, I can make a PR to make an `assert` instead of a `warn`, to be more explicit and to force the user to fix it as soon as this is done. What do you think?

**Comment by Turbo87** [comment]
I'm not sure. There might be cases where it's better to warn instead of throwing an exception and entirely breaking the app. On the other hand this code only runs in development mode, and not in production, so it might not be that bad 🤔 

@simplabs/devs any thoughts?

**Comment by marcoow** [comment]
On first glance it seems fine to make this an `assert` instead of a `warn`. I couldn't really think of a scenario where someone would want to define a `data-test-` attribute for a tagless component but not want that to break the app.

Of course this would have to be released in a major release though as it's a breaking change (at least in dev).

**Comment by mathieupoteriepeopledoc** [comment]
Thank you for your answers, @Turbo87 and @marcoow.
I just made the PR, I created a test in order to check if the error is thrown.
I also automatically ran prettier on the edited files (tell me if this is a problem).

I deleted the 'if' wrapping the previous code and set checked if `tagName !== ''` in the assert.

#205

**Comment by Turbo87** [comment]
resolved by #205

**Comment by samselikoff** [comment]
Seeing this for the first time as I'm updating our codebase, and we used to rely on this "feature" as explained here https://embermap.com/notes/92-ignoring-warnings.

tl;dr We use lots of tagless components but still want the interface for adding test ids to be the same:

```hbs
{{some-tagless-component data-test-id="foo"}}
```

The tagless component can use the `data-test-id` property internally and put it on the right place in its template. 

Thinking about a path forward for us given this change, I wanted to ask the team if y'all have thought about how these test selectors will work with Glimmer components, since I believe Glimmer components do/will work similarly to tagless components. That is, attributes that get added to a Glimmer component invocation don't automatically get applied anywhere; it's up to the component author to decide where to put `...attributes`. In this way, the approach we're taking with `data-test-id` on tagless Ember components seems to line up very closely with this.

Before this change we globally suppressed this warning since we were opting-in to this behavior. Any thoughts on perhaps adding an API, probably ideally at the component level, that would let users opt-in to allowing `data-test-id` to be passed through to tagless components?

**Comment by Turbo87** [comment]
@samselikoff if you use it as an attribute it should work fine for glimmer components, just not as a property

**Comment by samselikoff** [comment]
I see.

Since Ember components don't clearly distinguish, and it would be nice to use it as an "attribute" for Ember components, would you consider an API that lets us avoid or suppress this assertion?

**Comment by Turbo87** [comment]
```hbs
<SomeTaglessComponent data-test-id="foo" />
```

what currently happens when you try this?

**Comment by samselikoff** [comment]
It throws no error, and I can do this

```hbs
<button ...attributes>
  {{yield}}
</button>
```

in the component's template, and the `data-test-id` is forwarded correctly!

(Btw – woah!! This is so cool!)

So I guess the question is, can/do we want to "backport" this behavior to curly Ember components?

**Comment by Turbo87** [comment]
> I guess the question is, can/do we want to "backport" this behavior to curly Ember components?

IMHO it's not worth the extra complexity

**Comment by samselikoff** [comment]
Hm... seems a little strange to fork behavior based on whether one calls with angle brackets or curlies. Especially since both are going to be around for some time.

What about going back to the warning instead of an assertion, so users can opt-in to the behavior on their own?

**Comment by Turbo87** [comment]
> seems a little strange to fork behavior based on whether one calls with angle brackets or curlies

We're not forking behavior anywhere. With curlies, it's just impossible to use attribute bindings on tagless components, and we assume that using `data-test-foo` properties on them is a bug that we want to warn about.

Question is should we `warn()` (as we did before) or use a stronger warning like we do now. I wanted to suggest to pass `test-id` into your components instead but then realized that this would then not be removed by our AST transforms.

As you can see above @marcoow and I were both unsure if there are legitimate use-cases for keeping this a warning instead of an assertion. It seems that you have such use-cases, and I'm open to changing it back. The problem remains the same though, as you will always see the warning then, even if you're doing it deliberately. If we revert this we should think about a solution to actively suppress this warning for those cases.

**Comment by samselikoff** [comment]
Ah, I see. Thanks for explaining :)

Because Ember does have `registerWarnHandler` it's possible you could just delegate to that. That's what we were doing before and it worked well. You could add a note about it in the warning info/docs to let people know that's how they suppress the warning if they so desire.

**Comment by Turbo87** [comment]
that only suppresses globally and not per component though, right?

**Comment by samselikoff** [comment]
Correct. Was just thinking it might be nice to have a per-component API

**Comment by Turbo87** [comment]
We could add a `supportsDataTestProperties` property that a) will disable the `tagName == ''` check and b) will be removed by an AST transform so that it doesn't exist in prod builds. If we do that we wouldn't even need to revert the current behavior and could keep using assertions, and we could adjust the assertion message to hint about this workaround. PRs welcome 😉

**Comment by marcoow** [comment]
So I think there's a bunch of different things we're discussing here:

* Since non-Glimmer components don't provide any way to access the component's **outer** element from the template, the only way to add any attributes to that is via attribute bindings

```js
export default Component.extend({
});
```

* With Glimmer components, one can

**Comment by marcoow** [comment]
🤦‍♂️ looking at #291 I realize there's another case where you use a non-Glimmer component with `tagName` and still want to be able to pass `data-test-` etc. to forward those to child components. I guess we could change the assertion so that it checks whether the attributes are present on the prototype first before triggering the warning?

So if you explicitly defined the attributes you could opt-in to accepting `data-test-` attributes without triggering the warning, e.g.

```js
export default Component.extend({
  tagName: '',
  data-test-attr: null,
  …
});
```
